src/library_html5.js

```diff
@@ -1925,33 +2083,168 @@ var LibraryJSEvents = {
+    var canvas = JSEvents.findCanvasEventTarget(target);
+    if (!canvas) return {{{ cDefine('EMSCRIPTEN_RESULT_UNKNOWN_TARGET') }}};
+
+    if (canvas.canvasSharedPtr) {
+      // N.B. Reading the size of the Canvas takes priority from our shared state structure, which is not the actual size.
+      // However if is possible that there is a canvas size set event pending on an OffscreenCanvas owned by another thread,
+      // so that the real sizes of the canvas have not updated yet. Therefore reading the real values would be racy.
+      var w = {{{ makeGetValue('canvas.canvasSharedPtr', 0, 'i32') }}};
+      var h = {{{ makeGetValue('canvas.canvasSharedPtr', 4, 'i32') }}};
+      {{{ makeSetValue('width', 0, 'w', 'i32') }}};
+      {{{ makeSetValue('height', 0, 'h', 'i32') }}};
+    } else if (canvas.offscreenCanvas) {
+      {{{ makeSetValue('width', 0, 'canvas.offscreenCanvas.width', 'i32') }}};
+      {{{ makeSetValue('height', 0, 'canvas.offscreenCanvas.height', 'i32') }}};
+    } else if (!canvas.controlTransferredOffscreen) {
+      {{{ makeSetValue('width', 0, 'canvas.width', 'i32') }}};
+      {{{ makeSetValue('height', 0, 'canvas.height', 'i32') }}};
+    } else {
+#if GL_DEBUG
+      console.error('canvas.controlTransferredOffscreen but we do not own the canvas, and do not know who has (no canvas.canvasSharedPtr present, an internal bug?)!\n');
+#endif
+      return {{{ cDefine('EMSCRIPTEN_RESULT_UNKNOWN_TARGET') }}};
+    }
     return {{{ cDefine('EMSCRIPTEN_RESULT_SUCCESS') }}};
   },
 
-  emscripten_set_element_css_size: function(target, width, height) {
-    if (!target) {
-      target = Module['canvas'];
-    } else {
-      target = JSEvents.findEventTarget(target);
-    }
+  emscripten_get_canvas_element_size_main_thread__proxy: 'sync',
+  emscripten_get_canvas_element_size_main_thread__sig: 'iiii',
+  emscripten_get_canvas_element_size_main_thread__deps: ['emscripten_get_canvas_element_size_calling_thread'],
+  emscripten_get_canvas_element_size_main_thread: function(target, width, height) { return _emscripten_get_canvas_element_size_calling_thread(target, width, height); },
 
+  emscripten_get_canvas_element_size__deps: ['emscripten_get_canvas_element_size_calling_thread', 'emscripten_get_canvas_element_size_main_thread'],
+  emscripten_get_canvas_element_size: function(target, width, height) {
+    var canvas = JSEvents.findCanvasEventTarget(target);
+    if (canvas) return _emscripten_get_canvas_element_size_calling_thread(target, width, height);
+    else return _emscripten_get_canvas_element_size_main_thread(target, width, height);
+  }, 
+
+  // JavaScript-friendly API, returns pair [width, height]
+  $emscripten_get_canvas_element_size_js__deps: ['emscripten_get_canvas_element_size'],
+  $emscripten_get_canvas_element_size_js: function(target) {
+    var stackTop = Runtime.stackSave();
+    var w = Runtime.stackAlloc(8);
+    var h = w + 4;
```

Comment:
this will not work on the LLVM wasm backend, the stack grows down there.

`h = (w + stackTop) / 2` should work ;)

Reply:
Surely `stackAlloc()` will still return a pointer to the beginning of the allocated memory address on the stack? So if one calls `var p = stackAlloc(8)`, then `HEAPU8[p]` through `HEAPU8[p+7]` are the bytes that were allocated?

Reply:
Oh right, you add to the beginning of the allocation. That seems fine then.